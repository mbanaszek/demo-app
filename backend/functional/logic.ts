export const isNil = <T>(v: T | null | undefined): v is null | undefined => v === null || v === undefined;
export const isNotNil = <T>(v: T | null | undefined): v is T => v !== null && v !== undefined;
export const isTrue = (v: boolean): v is true => v === true;
export const isFalse = (v: boolean): v is true => v === false;