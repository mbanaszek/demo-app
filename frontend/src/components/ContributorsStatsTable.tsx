import {Table} from "react-bootstrap";
import React, {FunctionComponent} from "react";
import Image from 'react-bootstrap/Image'
import {ContributorStats} from "../fetchRepositoryContributorsStatistics";

interface ContributorsStatsTableProps {
    contributorsStats: ContributorStats[]
}

export const ContributorsStatsTable: FunctionComponent<ContributorsStatsTableProps> = ({contributorsStats}): JSX.Element => {

    const sortedContributedStats = sortContributorsStatsByHighestTotalChanges(contributorsStats);

    return (
        <Table striped bordered hover>
            <thead>
                <tr>
                    <th>#</th>
                    <th></th>
                    <th>Name</th>
                    <th></th>
                    <th></th>
                    <th className={"text-right"}>Total score</th>
                </tr>
            </thead>
            <tbody>
            {
                sortedContributedStats.length === 0
                ?
                    <tr>
                        <td colSpan={6}>Please provide a valid GitHub repository URL.</td>
                    </tr>
                : sortedContributedStats.map((contributorStats, index) => (
                    <tr key={contributorStats.name}>
                        <td>{index + 1}.</td>
                        <td>
                            <Image src={contributorStats.avatarUrl} thumbnail width={"150 px"}/>
                        </td>
                        <td>{contributorStats.name}</td>
                        <td className={"text-right"}>
                            <span className={"text-success"}>+ {contributorStats.additions} additions</span>
                        </td>
                        <td className={"text-right"}>
                            <span className={"text-danger"}>- {contributorStats.deletions} deletions</span>
                        </td>
                        <td className={"text-right"}>
                            <span className={"text-info"}>{contributorStats.additions + contributorStats.deletions} updates</span>
                        </td>
                    </tr>
                ))
            }
            </tbody>
        </Table>
    );
}

const sortContributorsStatsByHighestTotalChanges = (stats: ContributorStats[]): ContributorStats[] => {
    return stats.slice().sort((
        firstStats,
        secondStats
    ) => {
        const firstTotalChanges = firstStats.additions + firstStats.deletions;
        const secondTotalChanges = secondStats.additions + secondStats.deletions;
        return secondTotalChanges - firstTotalChanges;
    })
}

