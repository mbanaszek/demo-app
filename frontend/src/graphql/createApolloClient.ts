import { ApolloClient, HttpLink, InMemoryCache, NormalizedCacheObject } from "@apollo/client";

const graphqlAPIURL = process.env.NODE_ENV === "production" ? "https://boiling-fjord-26962.herokuapp.com/graphql" : "http://localhost:8001/graphql";

export const createApolloClient = (): ApolloClient<NormalizedCacheObject> => {
    return new ApolloClient({
        link: new HttpLink({ uri: graphqlAPIURL }),
        cache: new InMemoryCache(),
    });
};