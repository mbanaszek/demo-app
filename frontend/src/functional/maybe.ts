import { isNil } from "./logic";

export type Maybe<T> = T | null;

export const maybeWithDefault = <V>(defaultValue: V) => (maybe: Maybe<V>): V => (isNil(maybe) ? defaultValue : maybe);