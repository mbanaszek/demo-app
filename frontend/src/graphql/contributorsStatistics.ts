import {gql} from '@apollo/client';
import {Maybe} from "../functional/maybe";
import {isNil} from "../functional/logic";
const parseGithubURL = require('parse-github-url');

export type ContributorStats = {
    name: string;
    avatarUrl: string;
    additions: number;
    deletions: number;
};

export type RepositoryDetails = {
    owner: string;
    name: string;
}

export const getRepositoryDetails = (githubURL: string): Maybe<RepositoryDetails> => {

    const parsedGithubURL = parseGithubURL(githubURL);
    console.log(parsedGithubURL);

    if (isNil(parsedGithubURL) || isNil(parsedGithubURL.repository)) return null;

    const repositoryFullNameSplit = parsedGithubURL.repository.split('/');
    if (repositoryFullNameSplit.length !== 2) return null;

    return {
        owner: repositoryFullNameSplit[0],
        name: repositoryFullNameSplit[1]
    };
}

export const createContributorsStatisticsQuery = (repositoryDetails: RepositoryDetails): any => gql`
    {
      contributorsStats(
        repositoryOwner: "${repositoryDetails.owner}", 
        repositoryName: "${repositoryDetails.name}"
        ) {
            name
            avatarUrl
            additions
            deletions
        }
    }
`;