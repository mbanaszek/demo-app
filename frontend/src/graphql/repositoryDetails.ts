import { Maybe } from "../functional/maybe";
import parseGithubURL from "parse-github-url";
import { isNil } from "../functional/logic";

export type RepositoryDetails = {
    owner: string;
    name: string;
}

export const getRepositoryDetails = (githubURL: string): Maybe<RepositoryDetails> => {

    const parsedGithubURL = parseGithubURL(githubURL);

    if (isNil(parsedGithubURL) || isNil(parsedGithubURL.repository)) return null;

    const repositoryFullNameSplit = parsedGithubURL.repository.split("/");
    if (repositoryFullNameSplit.length !== 2) return null;

    return {
        owner: repositoryFullNameSplit[0],
        name: repositoryFullNameSplit[1]
    };
};