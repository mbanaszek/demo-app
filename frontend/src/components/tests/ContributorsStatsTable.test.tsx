import React from "react";
import { MockedProvider } from "@apollo/client/testing";
import TestRenderer, { act } from "react-test-renderer";

import { ContributorsStatsTable } from "../ContributorsStatsTable";
import {
    ContributorStats,
    createContributorsStatisticsQuery
} from "../../graphql/contributorsStatistics";
import { Maybe } from "../../functional/maybe";
import { isNil } from "../../functional/logic";
import { RepositoryDetails } from "../../graphql/repositoryDetails";

describe("Rendering table", () => {

    const repositoryDetails: RepositoryDetails = {
        name: "dummy-owner",
        owner: "dummy-name"
    };

    test("Loading.", async () => {
        // Given
        const apolloClientMock = createApolloMock(repositoryDetails, [], null, null);

        // When
        const component = renderTable(apolloClientMock, repositoryDetails);

        // Then
        await waitForRender();
        expect(component.toJSON()).toMatchSnapshot();
    });

    test("No results.", async () => {
        // Given
        const apolloClientMock = createApolloMock(repositoryDetails, [], null, null);

        // When
        const component = renderTable(apolloClientMock, repositoryDetails);

        // Then
        await waitForRender();
        expect(component.toJSON()).toMatchSnapshot();
    });

    test("Results.", async () => {
        // Given
        const dummyContributorsStats: ContributorStats[] = [
            { name: "dummy-name-1", avatarUrl: "http://dummy.avatar/url-1", additions: 1, deletions: 2 },
            { name: "dummy-name-2", avatarUrl: "http://dummy.avatar/url-2", additions: 3, deletions: 4 }
        ];
        const apolloClientMock = createApolloMock(repositoryDetails, dummyContributorsStats, null, null);

        // When
        const component = renderTable(apolloClientMock, repositoryDetails);

        // Then
        await waitForRender();
        expect(component.toJSON()).toMatchSnapshot();
    });

    test("Error message (like repository not found).", async () => {
        // Given
        const apolloClientMock = createApolloMock(
            repositoryDetails,
            null,
            [ { errorName: "RepositoryNotFound", statusCode: 404, message: "Repository not found." }],
            null
        );

        // When
        const component = renderTable(apolloClientMock, repositoryDetails);

        // Then
        await waitForRender();
        expect(component.toJSON()).toMatchSnapshot();
    });

    test("Network error message.", async () => {
        // Given
        const apolloClientMock = createApolloMock(
            repositoryDetails,
            null,
            null,
            "Network error."
        );

        // When
        const component = renderTable(apolloClientMock, repositoryDetails);

        // Then
        await waitForRender();
        expect(component.toJSON()).toMatchSnapshot();
    });
});

const createApolloMock = (repositoryDetails: RepositoryDetails, contributorsStats: Maybe<ContributorStats[]>, errors: Maybe<ApolloResponseErrors[]>, networkError: Maybe<string>): any => {
    return {
        request: {
            query: createContributorsStatisticsQuery(repositoryDetails)
        },
        ... isNil(errors) ? {} : { errors: errors },
        ... isNil(networkError) ? {} : { error: new Error(networkError) },
        result: {
            data: {
                contributorsStats: contributorsStats
            },
        },
    };
};

const renderTable = (apolloClientMock: any, repositoryDetails: RepositoryDetails) => TestRenderer.create(
    <MockedProvider mocks={[apolloClientMock]} addTypename={false}>
        <ContributorsStatsTable repositoryDetails={repositoryDetails} />
    </MockedProvider>,
);

const waitForRender = async () => act(() => {
    return new Promise(resolve => {
        setTimeout(resolve, 0);
    });
});

type ApolloResponseErrors = {
    "errorName": string;
    "statusCode": number;
    "message": string;
}