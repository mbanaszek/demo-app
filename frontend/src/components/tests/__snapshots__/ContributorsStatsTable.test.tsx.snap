// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering table Error message (like repository not found). 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th />
      <th>
        Name
      </th>
      <th />
      <th />
      <th
        className="text-right"
      >
        Total score
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={6}
      >
        Please provide a valid GitHub repository URL.
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Rendering table Loading. 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th />
      <th>
        Name
      </th>
      <th />
      <th />
      <th
        className="text-right"
      >
        Total score
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;

exports[`Rendering table Network error message. 1`] = `
<div
  className="text-center container"
>
  <div
    className="fade alert alert-danger show"
    role="alert"
  >
    Network problems.
  </div>
</div>
`;

exports[`Rendering table No results. 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th />
      <th>
        Name
      </th>
      <th />
      <th />
      <th
        className="text-right"
      >
        Total score
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;

exports[`Rendering table Results. 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th />
      <th>
        Name
      </th>
      <th />
      <th />
      <th
        className="text-right"
      >
        Total score
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
        .
      </td>
      <td>
        <img
          className="img-thumbnail"
          src="http://dummy.avatar/url-2"
          width="150 px"
        />
      </td>
      <td>
        dummy-name-2
      </td>
      <td
        className="text-right"
      >
        <span
          className="text-success"
        >
          + 
          3
           additions
        </span>
      </td>
      <td
        className="text-right"
      >
        <span
          className="text-danger"
        >
          - 
          4
           deletions
        </span>
      </td>
      <td
        className="text-right"
      >
        <span
          className="text-info"
        >
          7
           updates
        </span>
      </td>
    </tr>
    <tr>
      <td>
        2
        .
      </td>
      <td>
        <img
          className="img-thumbnail"
          src="http://dummy.avatar/url-1"
          width="150 px"
        />
      </td>
      <td>
        dummy-name-1
      </td>
      <td
        className="text-right"
      >
        <span
          className="text-success"
        >
          + 
          1
           additions
        </span>
      </td>
      <td
        className="text-right"
      >
        <span
          className="text-danger"
        >
          - 
          2
           deletions
        </span>
      </td>
      <td
        className="text-right"
      >
        <span
          className="text-info"
        >
          3
           updates
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
